http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/HonorCloseOnReleaseChannelPool.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import static software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils.doInEventLoop;
+
+import io.netty.channel.Channel;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Wrap a channel pool so that {@link ChannelAttributeKey#CLOSE_ON_RELEASE} is honored when a channel is released to the
+ * underlying pool.
+ *
+ * When a channel is released and {@link ChannelAttributeKey#CLOSE_ON_RELEASE} is true on the channel, the channel will be closed
+ * before it is released to the underlying pool.
+ */
+@SdkInternalApi
+public class HonorCloseOnReleaseChannelPool implements ChannelPool {
```

Comment:
Can we add some unit tests for this class? Let's add a test class for every new class :)

Reply:
I disagree. Too many class-level tests for simple classes with a large number of dependencies (like this one) makes the code base too rigid and hard to update. Class-level tests written for this class: (a) wouldn't verify it actually *works*, just that it calls the methods we think we should, and (b) would almost certainly have to be updated if we ever touch the implementation.

This class's behavior is better tested (imo) with the functional tests.

Reply:
Discussed in person. We agree that unit test class for `IdleConnectionReaperHandler.java ` is not needed since the functional tests would cover it. However, the functional test might not cover this class, so unit tests for this class are needed.  (Fun fight! 🍻)

Reply:
Added tests for the release-on-close behavior.